pkg/pgmodel/ingestor/metric_batcher.go

```diff
@@ -208,271 +202,21 @@ func (h *metricBatcher) flush() {
-	labelBatches := 0
-	for i := 0; i < len(names.Elements); i += itemsPerBatch {
-		labelBatches++
-		high := i + itemsPerBatch
-		if len(names.Elements) < high {
-			high = len(names.Elements)
-		}
-		namesSlice, err := names.Slice(i, high)
-		if err != nil {
-			return dbEpoch, 0, fmt.Errorf("Error filling labels: slicing names: %w", err)
-		}
-		valuesSlice, err := values.Slice(i, high)
-		if err != nil {
-			return dbEpoch, 0, fmt.Errorf("Error filling labels: slicing values: %w", err)
-		}
-		batch.Queue("BEGIN;")
-		batch.Queue("SELECT * FROM "+schema.Catalog+".get_or_create_label_ids($1, $2, $3)", metricName, namesSlice, valuesSlice)
-		batch.Queue("COMMIT;")
-	}
-	br, err := h.conn.SendBatch(context.Background(), batch)
-	if err != nil {
-		return dbEpoch, 0, fmt.Errorf("Error filling labels: %w", err)
-	}
-	defer br.Close()
-
-	if _, err := br.Exec(); err != nil {
-		return dbEpoch, 0, fmt.Errorf("Error filling labels on begin: %w", err)
-	}
-	err = br.QueryRow().Scan(&dbEpoch)
-	if err != nil {
-		return dbEpoch, 0, fmt.Errorf("Error filling labels: %w", err)
-	}
-	if _, err := br.Exec(); err != nil {
-		return dbEpoch, 0, fmt.Errorf("Error filling labels on commit: %w", err)
-	}
-
-	var count int
-	for i := 0; i < labelBatches; i++ {
-		if _, err := br.Exec(); err != nil {
-			return dbEpoch, 0, fmt.Errorf("Error filling labels on begin label batch: %w", err)
-		}
-
-		err := func() error {
-			var (
-				pos         []int32
-				labelIDs    []int32
-				labelNames  pgutf8str.TextArray
-				labelValues pgutf8str.TextArray
-				names       []string
-				values      []string
-			)
-			err := br.QueryRow().Scan(&pos, &labelIDs, &labelNames, &labelValues)
-			if err != nil {
-				return fmt.Errorf("Error filling labels: %w", err)
-			}
-			names = labelNames.Get().([]string)
-			values = labelValues.Get().([]string)
-
-			for i := range pos {
-				res := labelInfo{Pos: pos[i], labelID: labelIDs[i]}
-				key := labels.Label{Name: names[i], Value: values[i]}
-				_, ok := labelMap[key]
-				if !ok {
-					return fmt.Errorf("Error filling labels: getting a key never sent to the db")
-				}
-				labelMap[key] = res
-				if int(res.Pos) > maxPos {
-					maxPos = int(res.Pos)
-				}
-				count++
-			}
-			return nil
-		}()
-		if err != nil {
-			return dbEpoch, 0, err
-		}
-		if _, err := br.Exec(); err != nil {
-			return dbEpoch, 0, fmt.Errorf("Error filling labels on commit label batch: %w", err)
+	recvChannel := h.input
+	for {
+		if h.pending.IsFull() {
+			recvChannel = nil
 		}
-	}
-	if count != items {
-		return dbEpoch, 0, fmt.Errorf("Error filling labels: not filling as many items as expected: %v vs %v", count, items)
-	}
-	return dbEpoch, maxPos, nil
-}
+		numSeries := h.pending.batch.CountSeries()
+		select {
+		case h.toCopiers <- copyRequest{h.pending, h.metricTableName}:
+			MetricBatcherFlushSeries.Observe(float64(numSeries))
+			h.pending = NewPendingBuffer()
+			return
+		case req := <-recvChannel:
+			h.pending.addReq(req)
```

Comment:
Renamed the function. I am setting to nil to avoid receiving data. Setting channel to nil in select is idiomatic Go. https://www.godesignpatterns.com/2014/05/nil-channels-always-block.html